hathor/feature_activation/feature_service.py

```diff
@@ -0,0 +1,170 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import NamedTuple
+
+from hathor.feature_activation.feature import Feature
+from hathor.feature_activation.model.criteria import Criteria
+from hathor.feature_activation.model.feature_state import FeatureState
+from hathor.feature_activation.settings import Settings
+from hathor.transaction import Block
+
+
+class _Block(NamedTuple):
+    """A Block wrapper including the calculated height."""
+    data: Block
+    height: int
+
+    @classmethod
+    def from_block(cls, block: Block) -> '_Block':
+        return _Block(
+            data=block,
+            height=block.calculate_height()
+        )
+
+
+class FeatureService:
+    __slots__ = ('_settings',)
+
+    def __init__(self, *, settings: Settings) -> None:
+        self._settings = settings
+
+    def is_feature_active(self, *, block: Block, feature: Feature) -> bool:
+        """Returns whether a Feature is active at a certain block."""
+        state = self.get_state(block=block, feature=feature)
+
+        return state == FeatureState.ACTIVE
+
+    def get_state(self, *, block: Block, feature: Feature) -> FeatureState:
+        """Returns the state of a feature at a certain block."""
+
+        # per definition, the genesis block is in the DEFINED state for all features
+        if block.is_genesis:
+            return FeatureState.DEFINED
+
+        # All blocks within the same evaluation interval have the same state, that is, the state is only defined for
+        # the block in each interval boundary. Therefore, we get the state of the previous boundary.
+        _block = _Block.from_block(block)
+        offset_to_boundary = _block.height % self._settings.evaluation_interval
+        previous_boundary_height = _block.height - (offset_to_boundary or self._settings.evaluation_interval)
+        previous_boundary_block = _get_ancestor_at_height(_block=_block, height=previous_boundary_height)
+        previous_state = self.get_state(block=previous_boundary_block, feature=feature)
+
+        if offset_to_boundary != 0:
+            return previous_state
+
+        return self._calculate_new_state(
```

Comment:
If I got it right, this result of this calculation is immutable for each block. So, should we save it on block.metadata()?

Reply:
It is, that optimization is part of the Caching Mechanism described in [this section](https://github.com/HathorNetwork/rfcs/blob/master/projects/feature-activation/0001-feature-activation-for-blocks.md#caching-mechanism) of the design. It'll be implemented in a separate PR.

The only difference is that the design defines that only boundary blocks will persist this information, but it could be done for all blocks. I'll keep this in mind to check what's better when implementing it.

Reply:
Ok. It makes sense to keep it only on boundary blocks as you can easily calculate the previous boundary block height.